File: tests/test_genin_skill_contract.py
#!/usr/bin/env python3
import importlib.util
import json
import os
import sqlite3
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent


def load_migrate():
    spec = importlib.util.spec_from_file_location("konoha_migrate", ROOT / "src" / "migrate.py")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


class TestGeninSkillContract(unittest.TestCase):
    def test_shipped_skill_metadata_is_canonical(self):
        copies = [
            ROOT / "src" / "templates" / "skills" / "genin-skill",
            ROOT / ".agents" / "skills" / "genin-skill",
            ROOT / ".cursor" / "skills" / "genin-skill",
        ]
        for skill_dir in copies:
            self.assertTrue((skill_dir / "SKILL.md").is_file(), skill_dir)
            self.assertIn("name: genin-skill", (skill_dir / "SKILL.md").read_text(encoding="utf-8"))
            for metadata in (skill_dir / "agents").glob("*.yaml"):
                content = metadata.read_text(encoding="utf-8")
                self.assertIn("name: genin-skill", content, metadata)
                self.assertNotIn("name: deep-code-explorer", content, metadata)

        self.assertNotIn(
            "deep-code-explorer",
            (ROOT / ".agents" / "skills" / "kage-skill" / "SKILL.md").read_text(encoding="utf-8"),
        )
        self.assertNotIn(
            "deep-code-explorer",
            (ROOT / ".cursor" / "skills" / "kage-skill" / "SKILL.md").read_text(encoding="utf-8"),
        )

    def test_legacy_rows_canonicalize_with_canonical_precedence(self):
        migrate = load_migrate()
        with tempfile.TemporaryDirectory() as tmp:
            db_path = Path(tmp) / "skills.db"
            original = migrate.DB_PATH
            migrate.DB_PATH = str(db_path)
            try:
                conn = migrate.setup_db()
                conn.execute(
                    "INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)",
                    ("genin-skill", "genin-skill", "skill", "canonical"),
                )
                conn.execute(
                    "INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)",
                    ("deep-code-explorer", "deep-code-explorer", "skill", "legacy"),
                )
                conn.execute(
                    "INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)",
                    ("deep-code-explorer/code-review", "deep-code-explorer", "reference", "legacy ref"),
                )
                conn.commit()
                migrated, removed = migrate.normalize_legacy_skill_names(conn)
                conn.commit()
                self.assertEqual(migrated, 1)
                self.assertEqual(removed, 1)
                names = {row[0] for row in conn.execute("SELECT name FROM skills")}
                self.assertEqual(names, {"genin-skill", "genin-skill/code-review"})
                conn.close()
            finally:
                migrate.DB_PATH = original

    def test_clean_migration_requires_genin_skill(self):
        with tempfile.TemporaryDirectory() as tmp:
            skills_dir = Path(tmp) / "skills"
            skill_dir = skills_dir / "genin-skill"
            skill_dir.mkdir(parents=True)
            (skill_dir / "SKILL.md").write_text(
                "---\nname: genin-skill\ndescription: test\n---\n# Genin\n",
                encoding="utf-8",
            )
            db_path = Path(tmp) / "skills.db"
            result = subprocess.run(
                [
                    sys.executable,
                    str(ROOT / "src" / "migrate.py"),
                    "--clean",
                    "--skills-dir",
                    str(skills_dir),
                    "--skills",
                    "genin-skill",
                    "--db-path",
                    str(db_path),
                    "--require-skill",
                    "genin-skill",
                ],
                cwd=ROOT,
                capture_output=True,
                text=True,
                check=False,
            )
            self.assertEqual(result.returncode, 0, result.stderr + result.stdout)
            conn = sqlite3.connect(db_path)
            self.assertIsNotNone(conn.execute("SELECT 1 FROM skills WHERE name = 'genin-skill'").fetchone())
            self.assertEqual(
                conn.execute("SELECT COUNT(*) FROM skills WHERE name LIKE 'deep-code-explorer%'").fetchone()[0],
                0,
            )
            conn.close()


if __name__ == "__main__":
    unittest.main()

File: tests/test_genin_skill_contract.js
#!/usr/bin/env node
'use strict';

/**
 * tests/test_genin_skill_contract.js — Verifies Genin skill naming and migration contracts.
 */

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { execSync } = require('child_process');
const db = require('../src/db');
const migrate = require('../src/migrate');

const ROOT = path.resolve(__dirname, '..');

async function run() {
  console.log('Running test_genin_skill_contract tests...');

  // 1. Shipped skill metadata is canonical
  const copies = [
    path.join(ROOT, 'src', 'templates', 'skills', 'genin-skill'),
    path.join(ROOT, '.agents', 'skills', 'genin-skill'),
    path.join(ROOT, '.cursor', 'skills', 'genin-skill')
  ];
  for (const skillDir of copies) {
    const skillMd = path.join(skillDir, 'SKILL.md');
    assert.ok(fs.existsSync(skillMd), `${skillMd} must exist`);
    const content = fs.readFileSync(skillMd, 'utf-8');
    assert.ok(content.includes('name: genin-skill'), `name: genin-skill must be in ${skillMd}`);

    const agentsDir = path.join(skillDir, 'agents');
    if (fs.existsSync(agentsDir)) {
      const yamls = fs.readdirSync(agentsDir).filter(f => f.endsWith('.yaml'));
      for (const y of yamls) {
        const yContent = fs.readFileSync(path.join(agentsDir, y), 'utf-8');
        assert.ok(yContent.includes('name: genin-skill'), `name: genin-skill must be in ${y}`);
        assert.ok(!yContent.includes('name: deep-code-explorer'), `deep-code-explorer must not be in ${y}`);
      }
    }
  }

  const kageSkillMd1 = path.join(ROOT, '.agents', 'skills', 'kage-skill', 'SKILL.md');
  const kageSkillMd2 = path.join(ROOT, '.cursor', 'skills', 'kage-skill', 'SKILL.md');
  assert.ok(!fs.readFileSync(kageSkillMd1, 'utf-8').includes('deep-code-explorer'));
  assert.ok(!fs.readFileSync(kageSkillMd2, 'utf-8').includes('deep-code-explorer'));
  console.log('✓ Shipped skill metadata is canonical passed');

  // 2. Legacy rows canonicalize with canonical precedence
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'genin_contract_'));
  const testDbPath = path.join(tmpDir, 'skills.db');

  try {
    const conn = db.getDb(testDbPath);
    migrate.setupDb(conn);

    conn.prepare('INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)').run(
      'genin-skill', 'genin-skill', 'skill', 'canonical'
    );
    conn.prepare('INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)').run(
      'deep-code-explorer', 'deep-code-explorer', 'skill', 'legacy'
    );
    conn.prepare('INSERT INTO skills (name, skill_name, type, content) VALUES (?, ?, ?, ?)').run(
      'deep-code-explorer/code-review', 'deep-code-explorer', 'reference', 'legacy ref'
    );

    const { migrated, removed } = migrate.normalizeLegacySkillNames(conn);
    assert.strictEqual(migrated, 1);
    assert.strictEqual(removed, 1);

    const names = new Set(conn.prepare('SELECT name FROM skills').all().map(r => r.name));
    assert.deepStrictEqual(names, new Set(['genin-skill', 'genin-skill/code-review']));
    conn.close();
    console.log('✓ Legacy rows canonicalize with canonical precedence passed');

    // 3. Clean migration requires genin-skill
    const skillsDir = path.join(tmpDir, 'skills');
    const geninDir = path.join(skillsDir, 'genin-skill');
    fs.mkdirSync(geninDir, { recursive: true });
    fs.writeFileSync(path.join(geninDir, 'SKILL.md'), '---\nname: genin-skill\ndescription: test\n---\n# Genin\n');

    const migrateDbPath = path.join(tmpDir, 'clean_migrate.db');
    execSync(
      `node "${path.join(ROOT, 'src', 'migrate.js')}" --clean --skills-dir "${skillsDir}" --skills genin-skill --db-path "${migrateDbPath}" --require-skill genin-skill --skip-embeddings`,
      { cwd: ROOT, stdio: 'pipe' }
    );

    const conn2 = db.getDb(migrateDbPath);
    const hasGenin = conn2.prepare("SELECT 1 FROM skills WHERE name = 'genin-skill'").get();
    assert.ok(hasGenin, 'genin-skill must exist in clean migration db');

    const legacyCount = conn2.prepare("SELECT COUNT(*) as count FROM skills WHERE name LIKE 'deep-code-explorer%'").get().count;
    assert.strictEqual(legacyCount, 0, 'No deep-code-explorer rows should exist');
    conn2.close();
    console.log('✓ Clean migration requires genin-skill passed');

    console.log('\nAll test_genin_skill_contract tests passed!');
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

if (require.main === module) {
  run().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { run };
